tests/integration_test/backend_retry_test.go

```diff
@@ -0,0 +1,212 @@
+			wantSpanNames: []string{
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"ingress Echo",
+			},
+		},
+		{
+			desc:                       "Failure, the retryNum is not enough",
+			backendRejectRequestStatus: 503,
+			backendRejectRequestNum:    2,
+			backendRetryOnsFlag:        "5xx",
+			backendRetryNumFlag:        1,
+			wantError:                  "503 Service Unavailable",
+			wantSpanNames: []string{
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"ingress Echo",
+			},
+		},
+		{
+			desc:                       "Failure, invalid retryOns",
+			backendRejectRequestStatus: 503,
+			backendRejectRequestNum:    2,
+			backendRetryOnsFlag:        "this-is-random-retryOn",
+			backendRetryNumFlag:        2,
+			wantError:                  "503 Service Unavailable",
+			wantSpanNames: []string{
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"ingress Echo",
+			},
+		},
+		{
+			desc:                       "Failure, no retry as 5xx doesn't cover 403",
+			backendRejectRequestStatus: 403,
+			backendRejectRequestNum:    2,
+			backendRetryOnsFlag:        "5xx",
+			backendRetryNumFlag:        2,
+			wantError:                  "403 Forbidden",
+			wantSpanNames: []string{
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"ingress Echo",
+			},
+		},
+		{
+			desc:                       "Success, sufficient retryNum and covered retryOn",
+			backendRejectRequestStatus: 503,
+			backendRejectRequestNum:    2,
+			backendRetryOnsFlag:        "5xx",
+			backendRetryNumFlag:        2,
+			wantResp:                   `{"message":""}`,
+			wantSpanNames: []string{
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"router backend-cluster-echo-api.endpoints.cloudesf-testing.cloud.goog_local egress",
+				"ingress Echo",
+			},
+		},
+	}
+	for _, tc := range testData {
+		func() {
+			configId := "test-config-id"
+			args := []string{
+				"--service_config_id=" + configId,
+				"--rollout_strategy=fixed",
+				"--suppress_envoy_headers",
+			}
+			if tc.backendRetryOnsFlag != defaultBackendRetryOns {
+				args = append(args, "--backend_retry_ons="+tc.backendRetryOnsFlag)
+			}
+
+			if tc.backendRetryNumFlag != defaultBackendRetryNum {
+				args = append(args, fmt.Sprintf("--backend_retry_num=%v", tc.backendRetryNumFlag))
+			}
+
+			s := env.NewTestEnv(platform.TestBackendRetry, platform.EchoSidecar)
```

Comment:
Do you think we should have a test for dynamic routing backends too, since remote backends is the primary use case where we need retries?

I assume the logic should be the same, so probably doesn't matter.

Reply:
Done.